src/manage_sql_secinfo.c

```diff
@@ -3601,285 +3607,843 @@ set_root_id (long int id, long int root_id)
+                               db_copy_buffer_t *cpe_match_nodes_buffer)
+{
+  if (secinfo_fast_init)
+    {
+      gchar *escaped_operator = sql_copy_escape (node_operator);
+
+      (*node_id_ptr)++;
+      if (*root_id_ptr < 0)
+        *root_id_ptr = *node_id_ptr;
+
+      if (db_copy_buffer_append_printf (cpe_match_nodes_buffer,
+                                        "%llu\t%llu\t%llu\t%s\t%d\n",
+                                        *node_id_ptr,
+                                        *root_id_ptr,
+                                        cve_db_id,
+                                        escaped_operator,
+                                        negate))
+        {
+          g_free (escaped_operator);
+          return -1;
+        }
+      g_free (escaped_operator);
+    }
+  else
+    {
+      *node_id_ptr = save_node (cve_db_id, node_operator, negate);
+
+      if (*node_id_ptr < 0)
+        *node_id_ptr = *node_id_ptr;
+
+      set_root_id (*node_id_ptr, *root_id_ptr);
+    }
+
+  return 0;
+}
+
+/**
+ * @brief Add an affected product and its placeholder CPE to the database
+ *
+ * This uses COPY SQL buffers. Methods not using them have to add affected
+ *  products a different way.
+ *
+ * @param[in]     cve_db_id       Database rowid of the CVE being processed
+ * @param[in,out] cpe_db_id_ptr   Pointer to DB rowid for placeholder CPEs
+ * @param[in]  match_cpe_name     CPE name of the product
+ * @param[in]  match_cpe_name_id  CPE name ID of the product
+ * @param[in]  hashed_cpes        Hashtable of CPE-IDs in the database
+ * @param[in]  current_cve_cpes   Hashtable of CPE rowids for current CVE
+ * @param[in]  cve_cpes_copy_buffer            COPY buffer for placeholder CPEs
+ * @param[in]  affected_products_copy_buffer   COPY buffer for affected products
+ */
+static int
+handle_cve_affected_product (resource_t cve_db_id,
+                             resource_t *cpe_db_id_ptr,
+                             const char *match_cpe_name,
+                             const char *match_cpe_name_id,
+                             GHashTable *hashed_cpes,
+                             GHashTable *current_cve_cpes,
+                             db_copy_buffer_t *cve_cpes_copy_buffer,
+                             db_copy_buffer_t *affected_products_copy_buffer)
+{
+  resource_t affected_product_cpe_db_id;
+
+  affected_product_cpe_db_id = GPOINTER_TO_INT (
+    g_hash_table_lookup (hashed_cpes, match_cpe_name)
+  );
+  if (g_hash_table_contains (hashed_cpes, match_cpe_name) == 0)
+    {
+      (*cpe_db_id_ptr)++;
+      affected_product_cpe_db_id = (*cpe_db_id_ptr);
+      gchar *escaped_cpe_name, *escaped_cpe_name_id;
+
+      escaped_cpe_name = sql_copy_escape (match_cpe_name);
+      escaped_cpe_name_id = sql_copy_escape (match_cpe_name_id);
+
+      g_hash_table_insert (hashed_cpes,
+                            g_strdup (match_cpe_name),
+                            GINT_TO_POINTER (*cpe_db_id_ptr));
```

Comment:
The indentation might need a fix here.
```suggestion
                           g_strdup (match_cpe_name),
                           GINT_TO_POINTER (*cpe_db_id_ptr));
```